packages/core/src/render3/di_setup.ts

```diff
@@ -179,6 +179,17 @@ function resolveProvider(
   }
 }
 
+/**
+ * Function is used as a destroy hook for Factory Providers. This function is invoked in a context
+ * of a provider instance, checks whether there is an `ngOnDestroy` function and invokes it if it's
+ * present.
+ */
+function ngOnDestroyForFactoryProviders(this: any) {
+  if (this && this.ngOnDestroy) {
+    this.ngOnDestroy();
+  }
```

Comment:
As `this.ngOnDestroy` is a megamorphic read, consider reading it only once:

```suggestion
  const onDestroy = this && this.ngOnDestroy;
  if (onDestroy) {
    onDestroy.call(this);
  }
```

Reply:
Can you provide me some performance tests?

just kidding lol, thanks for pointing it out:)

Reply:
😅